opt: hoist WITHs to the top level

justinj opened the pull request:
This commit changes how we build WITHs to always hoist them up to the
highest level possible. In most cases this means they're hoisted up to
be the root expression of the tree, but in some cases they cannot travel
further up past certain operators which logically denote a different
query (say, the input to an EXPLAIN).

It also stops WITHs from being correlated, fixing an overlooked case
from before.

This will mean that pruning columns through WITH expressions is
increasingly important and I will work on that next.

Release note (sql change): WITH expressions can no longer be correlated,
even when using LATERAL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41033)
<!-- Reviewable:end -->

justinj commented on pkg/sql/opt/optbuilder/select.go:
```diff
@@ -565,12 +565,13 @@ func (b *Builder) buildWithOrdinality(colName string, inScope *scope) (outScope
 	outScope = inScope.push()
 
-	start := len(b.ctes)
+	// Make a fake subquery to ensure that no CTEs are correlated.
```
@rytaft is this correct/ok to do?

RaduBerinde reviewed (commented):
I thought we only wanted to pull up `With`s that have mutations / side-effects?

FWIW postgres supports correlated `With` expressions when they're not mutations (and mutations inside subqueries error out with `WITH clause containing a data-modifying statement must be at the top level`). We could do the same (but we still want to hoist up the implicit Withs created by `[ .. ]`).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LpYz7mT3e1BvFaOzIRp:b-uvg1g9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1838 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpZf0lRAtZfSbf9n0H6:-LpZf0lSCgLp4bhPhi63:bgf2hun) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/norm/custom_funcs.go#L1838)):*
> ```Go
> 	}
> 
> 	if w, ok := r.(*memo.WithExpr); ok {
> ```

A comment would help here
___
*[pkg/sql/opt/norm/testdata/rules/decorrelate, line 2103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpZfD9C4yHQzFohBRSr:-LpZfD9C4yHQzFohBRSs:bb85618) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/norm/testdata/rules/decorrelate#L2103)):*
> ```
>     FAMILY "primary" (_int2, _int8, _timestamptz, _bool, _decimal, _string, rowid)
> )
> ----
> ```

This table can be deleted. Not sure if there is another way to add a regression test for #40592...
___
*[pkg/sql/opt/optbuilder/select.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpYwV_9lLbk12-Pcvgm-r1-571:-LpZgNhsACF6CaiqFzsC:bst53u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/optbuilder/select.go#L571)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

@rytaft is this correct/ok to do?
</blockquote></details>

I think this probably works, but it's sort of hard to reason about. Another option might be: whenever you build a CTE, just pass in an empty scope (similar to how Explain works). 

If you want to support correlated CTEs that don't have mutations, however, then maybe creating a fake subquery is the way to go....


<!-- Sent from Reviewable.io -->

Commit:
opt: block mutating non-top-level CTEs

This change simplifies hoisting CTEs quite a bit.

Release note (sql change): Mutations in CTEs not at the top level are no
longer allowed. This restriction is part of Postgres as well.

Commit:
opt: turn [...] into syntactic sugar for WITH

Release note: None

justinj reviewed (commented):
Ok, I fleshed this out a bunch more and addressed comments, I probably shouldn't have made it so much bigger, but that's where we are :/ I can break it up into smaller PRs if that would make things easier, it's already in three commits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LrFR1MYD3Y-EaA6Ahi2:br8lv10)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1838 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpZf0lRAtZfSbf9n0H6:-Lqg-lGJ76svXp2V9i0W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/norm/custom_funcs.go#L1838)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

A comment would help here
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/decorrelate, line 2103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpZfD9C4yHQzFohBRSr:-LqgEPG685ARqQEkK2C2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/norm/testdata/rules/decorrelate#L2103)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This table can be deleted. Not sure if there is another way to add a regression test for #40592...
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpYwV_9lLbk12-Pcvgm-r1-571:-LrAQXoX6cjHWOLikzgc:b-xsvddh) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/optbuilder/select.go#L571)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this probably works, but it's sort of hard to reason about. Another option might be: whenever you build a CTE, just pass in an empty scope (similar to how Explain works). 

If you want to support correlated CTEs that don't have mutations, however, then maybe creating a fake subquery is the way to go....
</blockquote></details>

I think we can't just pass in an empty scope, since CTEs still have to be able to refer to parent CTEs.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Very cool! Especially like how the second commit turned out!

On the first commit: we are adding this new `buildCtx` but we don't seem to be threading it everywhere, and in a lot of places we assume that the empty context is ok to pass down (which won't be true if we want to expand the use of this context; there should probably be a `child()` method on the input context). Until now, we've been using the `scope` to store everything, couldn't we add the `atRoot` flag there?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LrG0BUu7uI1jC_vNVrf:b-imkjrg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/optbuilder/select.go, line 580 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1Y_62JVEXuTTlV2R:-LrG1Y_62JVEXuTTlV2S:bke93s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e436512ed73f6ba89dca70d443b0bb9dee00e58b/pkg/sql/opt/optbuilder/select.go#L580)):*
> ```Go
> 	outScope = inScope.push()
> 
> 	// Make a fake subquery to ensure that no CTEs are correlated.
> ```

I thought we would still allow correlated CTEs that aren't mutations? (and only pull up mutations, and all square bracket stuff). Restricting read-only CTEs looks like it could be a regression in functionality (but no idea if it matters). Sorry if we discussed this and I forgot the latest conclusion
___
*[pkg/sql/opt/optbuilder/select.go, line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG2ljG88OAXv-8ptp2:-LrG2ljG88OAXv-8ptp3:bhir38r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/optbuilder/select.go#L173)):*
> ```Go
> 	inScope.cols = nil
> 	i := 0
> 	for _, col := range cte.cols {
> ```

[nit] looks like we can just do `for i, col :=`
___
*[pkg/sql/opt/xform/optimizer.go, line 791 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1wEC-xFTjX9vt5Kf:-LrG1wEC-xFTjX9vt5Kg:b5v2la4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/optimizer.go#L791)):*
> ```Go
> 	neededCols := rootProps.ColSet()
> 	if !neededCols.SubsetOf(root.Relational().OutputCols) {
> 		fmt.Println("yeet")
> ```

needs cleanup


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice!

I agree with Radu that it feels like `atRoot` should be part of `scope`.

> Reviewed 17 of 19 files at r2, 13 of 14 files at r3, 23 of 23 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LrL5-aR4hbLxCAucyel:b-m285lz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)

___
*[pkg/sql/opt/norm/testdata/rules/ordering, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrKzjMa5vRKcg_WjiGg:-LrKzjMa5vRKcg_WjiGh:b-29q6ta) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/norm/testdata/rules/ordering#L221)):*
> ```
>  │         └── (123,) [type=tuple{int}]
>  └── projections
>       └── variable: field [type=string, outer=(3)]
> ```

Why can't this be a passthrough column anymore?
___
*[pkg/sql/opt/optbuilder/select.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpYwV_9lLbk12-Pcvgm-r1-571:-LrL46-X3kwGmqqsw7wK:b-vpu4eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/optbuilder/select.go#L571)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think we can't just pass in an empty scope, since CTEs still have to be able to refer to parent CTEs.
</blockquote></details>

Ok! I can't think of any reason why this would fail...

But to be safe, have you tested something like including a WITH inside another correlated subquery?
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrL2IaC36xcY-3rTKDQ:-LrL2IaC36xcY-3rTKDR:b-xu1blh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/testdata/physprops/ordering#L1163)):*
> ```
>  ├── columns: a:7(int!null) b:8(int!null) c:9(int!null)
>  ├── ordering: +8
>  └── with &1
> ```

shouldn't WITHs that are only scanned once be optimized away?
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 513 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrL2oLW7ws9x8j4is54:-LrL2oLW7ws9x8j4is55:b-cwvn6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/testdata/rules/join_order#L513)):*
> ```
>     JOIN [INSERT INTO x (a) SELECT NULL FROM x RETURNING 1] ON false
>     JOIN x ON true
>     JOIN [UPDATE x SET a = 1 RETURNING 1] ON true
> ```

Is this test still testing what it was meant to?


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
I didn't thread it through because it seemed difficult to verify it was threaded everywhere, and would be difficult to test that that remained correct. In retrospect this has the same problem. Going to think about how this could be improved.

The problem with using the `scope` is that we use the same scope object in situations that change the build context (say, a subselect used as a data source). We could perhaps set a "context" on the scope and `defer` it back to what it was before, but that seems kind of sketchy and I think it would probably be more reasonable to just have the two concepts separate. Maybe the way to phrase this is that scope boundaries and context boundaries are different? Or maybe there's a way they can be unified somehow? It's not obvious to me how to do that. It does seem problematic to have these as two separate concepts, I'm open to being convinced there's a better solution here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LrKiynw48QUVbeA01v9:b9epub0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/norm/testdata/rules/ordering, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrKzjMa5vRKcg_WjiGg:-LrPOPp8-2NwcwyYYn4T:b-id9tsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/norm/testdata/rules/ordering#L221)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why can't this be a passthrough column anymore?
</blockquote></details>

I think because we hoist it, then subsequently inline it, rather than building it inlined. I'm not sure if there's a nice way to fix it short of having some kind of norm rule to re-number columns that could be passthrough columns after a normalization occurred?
___
*[pkg/sql/opt/optbuilder/select.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LpYwV_9lLbk12-Pcvgm-r1-571:-LrPdohH4z4kGtclulOQ:b-s0ro35) ([raw file](https://github.com/cockroachdb/cockroach/blob/453e2c6b945c3ccee25330b2147bda7459785b38/pkg/sql/opt/optbuilder/select.go#L571)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Ok! I can't think of any reason why this would fail...

But to be safe, have you tested something like including a WITH inside another correlated subquery?
</blockquote></details>

Added a test for this.
___
*[pkg/sql/opt/optbuilder/select.go, line 580 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1Y_62JVEXuTTlV2R:-LroSXZT6QvA8h3UeyNO:b-ivekh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e436512ed73f6ba89dca70d443b0bb9dee00e58b/pkg/sql/opt/optbuilder/select.go#L580)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I thought we would still allow correlated CTEs that aren't mutations? (and only pull up mutations, and all square bracket stuff). Restricting read-only CTEs looks like it could be a regression in functionality (but no idea if it matters). Sorry if we discussed this and I forgot the latest conclusion
</blockquote></details>

I think making this change is doable here but I'd rather make it a follow-up, just because there's some hairiness (it introduces two separate code-paths for where WITHs are introduced and I'm finding it tricky to square it with the rest of the changes here.
___
*[pkg/sql/opt/optbuilder/select.go, line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG2ljG88OAXv-8ptp2:-LrKgb-dCR0_PiXykPvp:b-nrea6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/optbuilder/select.go#L173)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] looks like we can just do `for i, col :=`
</blockquote></details>

Hm, not sure why I had it like this before!
___
*[pkg/sql/opt/xform/optimizer.go, line 791 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1wEC-xFTjX9vt5Kf:-LrKiur794b_x3qvqxna:b-i6dqh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/optimizer.go#L791)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

needs cleanup
</blockquote></details>

Yikes, fixed
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrL2IaC36xcY-3rTKDQ:-LrjNRnOEo-n6Z9OVyIk:bh0tjh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/testdata/physprops/ordering#L1163)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

shouldn't WITHs that are only scanned once be optimized away?
</blockquote></details>

We don't do so in the presence of side effects. We could maybe do it in some limited cases but it's by design that we want to hoist up all the mutations.
___
*[pkg/sql/opt/xform/testdata/rules/join_order, line 513 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrL2oLW7ws9x8j4is54:-LrjOCDo1XYB_53J_sfz:b-101cz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/testdata/rules/join_order#L513)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this test still testing what it was meant to?
</blockquote></details>

This change actually obviates what this test was originally supposed to test, so deleted!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
I think I still don't fully understand the difference between scope and context... is there some formal definition of when the context changes?

Otherwise, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 43 files at r5, 12 of 14 files at r6, 23 of 23 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LrunP9X5Ayn1BPz4dQE:b-eh7bge)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj and @RaduBerinde)

___
*[pkg/sql/opt/norm/testdata/rules/ordering, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrKzjMa5vRKcg_WjiGg:-LrukNLgDWfHrLEvVwFo:bk7syna) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/norm/testdata/rules/ordering#L221)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think because we hoist it, then subsequently inline it, rather than building it inlined. I'm not sure if there's a nice way to fix it short of having some kind of norm rule to re-number columns that could be passthrough columns after a normalization occurred?
</blockquote></details>

Makes sense. Probably not worth trying to fix at this point...
___
*[pkg/sql/opt/optbuilder/misc_statements.go, line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrulWXR55Aqo9F1lGDo:-LrulWXR55Aqo9F1lGDp:b-o9ntas) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b4ff216d35f76058dec0c0c4bb409e6640d737/pkg/sql/opt/optbuilder/misc_statements.go#L59)):*
> ```Go
> 	emptyScope := &scope{builder: b}
> 	colTypes := []*types.T{types.String}
> 	inputScope := b.buildStmt(n.Queries, colTypes, emptyScope, ctx)
> ```

why do some of these have `ctx.child()` but others just have `ctx`?
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 1163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrL2IaC36xcY-3rTKDQ:-LrumlTz7t5YBD_iu4Yr:bjegzgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3dac9b2d699346380cf0f217eb04e93bdd7ee0/pkg/sql/opt/xform/testdata/physprops/ordering#L1163)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

We don't do so in the presence of side effects. We could maybe do it in some limited cases but it's by design that we want to hoist up all the mutations.
</blockquote></details>

Ok! Thanks for the explanation


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
The only definition I have is "when it needs to change", whereas scopes have a semantically-defined lifecycle (right? or else hopping up the scope tree to figure out the right level for an aggregation to live at wouldn't make sense). If you think there's a way to mangle this into the scope I'd be happy to discuss! I'm not super happy with this as it's ended up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LryfJDd9TYTq9D4T0S_:b-btjrwp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj and @RaduBerinde)



<!-- Sent from Reviewable.io -->

Commit:
opt: hoist WITH expressions up to the top level

Release note (sql change): WITH expressions are now hoisted to the top
level in a query when possible.

justinj reviewed (commented):
Ok, finally rebased—PTAL! Sorry for how long this is taking—lots of finicky `scope` stuff to work out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-Lta6CJK61GJZ73-w3qv:b-nj4zpk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
Looks great to me, just some minor comments!

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 1 of 43 files at r5, 10 of 44 files at r8, 10 of 12 files at r9, 24 of 24 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-LteO59fD4YBOSXJT6mY:b24xdqj)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/delete.go, line 44 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteIxLnAnyNtV6FHK_7:-LteIxLnAnyNtV6FHK_8:bkpsyxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db2ec4c64cdc48b62080852604b9c7754c347f5/pkg/sql/opt/optbuilder/delete.go#L44)):*
> ```Go
> 	var wrapWiths func(*scope)
> 	inScope, wrapWiths = b.processWith(del.With, inScope)
> 	defer func() { wrapWiths(outScope) }()
> ```

Why not just call `wrapWiths()` before `return`? The magic defer that modifies the result is usually trickier to think about.

We could also consider taking the `processWith` call outside of these functions and change it to take a function for building the specific statement. Something like
```go
func (b *Builder) processWith(with *tree.With, inScope *scope, buildStmt fn(inScope *scope) *scope) *scope {
  ...
  inScope = b.buildCTEs(with, inScope)
  outScope := buildStmt(inScope)
  ...
  return outScope
}
```
In most cases the builder can just call this directly, e.g.
```go
  return b.processWith(ins.With, inScope, func (inScope *scope) *scope {
    return b.buildInsert(ins, inScope)
  })
```
We'd split `buildSelect` to have a wrapper that does this since that is used in a bunch of places.
___
*[pkg/sql/opt/optbuilder/scope.go, line 286 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteNZKl6pPa2VgQbXgG:-LteNZKm7eCDCJUbFYQW:bif5sj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/408476b7303ab457550833b02f53e451b441bf64/pkg/sql/opt/optbuilder/scope.go#L286)):*
> ```Go
> }
> 
> func (s *scope) makePresentationWithHiddenCols() physical.Presentation {
> ```

[nit] add a comment, this sounds like is should be used only in this special case
___
*[pkg/sql/opt/optbuilder/select.go, line 580 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1Y_62JVEXuTTlV2R:-LteLxAN-vSlpHawhuR6:b-k0hhx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e436512ed73f6ba89dca70d443b0bb9dee00e58b/pkg/sql/opt/optbuilder/select.go#L580)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think making this change is doable here but I'd rather make it a follow-up, just because there's some hairiness (it introduces two separate code-paths for where WITHs are introduced and I'm finding it tricky to square it with the rest of the changes here.
</blockquote></details>

Sure, just file an issue so we don't forget
___
*[pkg/sql/opt/optbuilder/select.go, line 583 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteMm5n5qIQ088JLJa3:-LteMm5n5qIQ088JLJa4:b5rd6bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L583)):*
> ```Go
> 	addedCTEs := make([]cteSource, len(with.CTEList))
> 
> 	// Make a fake subquery to ensure that no CTEs are correlated.
> ```

Add a TODO here if we plan to relax this
___
*[pkg/sql/opt/optbuilder/select.go, line 634 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteLhHa4-OXHxGKvRas:-LteLhHa4-OXHxGKvRat:bbobznw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L634)):*
> ```Go
> }
> 
> // constructed within an EXPLAIN should not be hoisted above it, and so we need
> ```

Missing text
___
*[pkg/sql/opt/optbuilder/select.go, line 648 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteMRbI4bn-bVusFB4i:-LteMRbI4bn-bVusFB4j:b-clvmi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L648)):*
> ```Go
> // flushCTEs adds With expressions on top of an expression.
> func (b *Builder) flushCTEs(expr memo.RelExpr) memo.RelExpr {
> 	var ctes []cteSource
> ```

[nit] `ctes := b.cteStack[len(b.cteStack)-1]` on the first line would be easier to read


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
TFTRs! Gonna finally try to merge this pupper

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-Lu3MT4xCUsWRtI1bKIw:b-l4plia)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/delete.go, line 44 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteIxLnAnyNtV6FHK_7:-Ltf5jNU66DxaGgslnkD:b-51453j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db2ec4c64cdc48b62080852604b9c7754c347f5/pkg/sql/opt/optbuilder/delete.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why not just call `wrapWiths()` before `return`? The magic defer that modifies the result is usually trickier to think about.

We could also consider taking the `processWith` call outside of these functions and change it to take a function for building the specific statement. Something like
```go
func (b *Builder) processWith(with *tree.With, inScope *scope, buildStmt fn(inScope *scope) *scope) *scope {
  ...
  inScope = b.buildCTEs(with, inScope)
  outScope := buildStmt(inScope)
  ...
  return outScope
}
```
In most cases the builder can just call this directly, e.g.
```go
  return b.processWith(ins.With, inScope, func (inScope *scope) *scope {
    return b.buildInsert(ins, inScope)
  })
```
We'd split `buildSelect` to have a wrapper that does this since that is used in a bunch of places.
</blockquote></details>

Done, yeah my apprehension for a while was that I wanted it to still be safe to just call `buildSelect`, but after making this change I do think it's a lot better.
___
*[pkg/sql/opt/optbuilder/misc_statements.go, line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrulWXR55Aqo9F1lGDo:-LtjzSRp-kJVLwOSdW9C:b-xbq292) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b4ff216d35f76058dec0c0c4bb409e6640d737/pkg/sql/opt/optbuilder/misc_statements.go#L59)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why do some of these have `ctx.child()` but others just have `ctx`?
</blockquote></details>

Not relevant any more
___
*[pkg/sql/opt/optbuilder/scope.go, line 286 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteNZKl6pPa2VgQbXgG:-LtzENyY6zjbPHQGBYms:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/408476b7303ab457550833b02f53e451b441bf64/pkg/sql/opt/optbuilder/scope.go#L286)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment, this sounds like is should be used only in this special case
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 580 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LrG1Y_62JVEXuTTlV2R:-LtzHmKh-_9qMYxLvlVo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e436512ed73f6ba89dca70d443b0bb9dee00e58b/pkg/sql/opt/optbuilder/select.go#L580)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sure, just file an issue so we don't forget
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 583 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteMm5n5qIQ088JLJa3:-LtzHmvg8_-ykk0aU7ng:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L583)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a TODO here if we plan to relax this
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 634 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteLhHa4-OXHxGKvRas:-LtzHuLn2yOxkl2SFHRr:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L634)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Missing text
</blockquote></details>

Fixed.
___
*[pkg/sql/opt/optbuilder/select.go, line 648 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-LteMRbI4bn-bVusFB4i:-LtzI4G2A25m-AA8FcfM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3862beb8419c4dd2a6ac11367ae08187f2eb4f3/pkg/sql/opt/optbuilder/select.go#L648)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `ctes := b.cteStack[len(b.cteStack)-1]` on the first line would be easier to read
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
rebased, trying again

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41033#-:-Lu3l9tX4fiH8tJlyFQY:b-l8v54b)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

justinj commented:
looked like a flake, givin it another go

bors r+
